`jj git push`: safety checks when pushing branch sideways, similar to `git push --force-with-lease`

ilyagr opened the pull request:
Previously, `jj git push` only made sure that the branch is in the expected
  location on the remote server when pushing a branch forward (as opposed to
  sideways or backwards). Now, `jj git push` makes a safety check in all cases
  and fails whenever `jj git fetch` would have introduced a conflict.

In other words, previously branches that moved sideways or backward were
  pushed similarly to Git's `git push --force`; now they have protections
  similar to `git push --force-with-lease` (though not identical to it, to match
  the behavior of `jj git fetch`). Note also that because of the way `jj git
  fetch` works, `jj` does not suffer from the same problems as Git's `git push
  --force-with-lease` in situations when `git fetch` is run in the background.

-------------------------

There are a few TODOs that may be added later (in this or separate PRs):

- [ ] Document this behavior in `jj help git push`, possibly elsewhere.
- [x] Adjustments to user-facing UI
- [ ] Additional tests

See TODOs in the code for more details. It's possible I forgot some, but it's probably time for others to look at the general form of this PR in any case.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- (n/a) I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

necauqua commented on lib/src/git.rs:
```diff
@@ -1305,53 +1317,110 @@ pub fn push_updates(
+                    // TODO: Suggestions to remove the double clone are welcome.
+                    // First, Rust lulls you into a sense of security. Then, out
+                    // of nowhere, an hour is spent looking for an idiomatic way
+                    // of converting &&Option<CommitId> to Option<CommitId>.
```
just `(**expr).clone()`?
I mean I guess it's also a bit ugly but that's what comes to mind when you ask for `&&Option<T> -> Option<T>`, and it's not worse than two clones

ilyagr replied:
Are you sure that works. IIRC, it did not.

necauqua replied:
Since clone takes &self, it would autoref to `&**expr`, `*x` does not take ownership unconditionally (I had a similar intuition myself until encountering similar things fixed it for me).

I did try it locally before commenting to be sure, but just with rust-analyzer actually, let me `cargo build` it.. - seems to work ¯\\\_(ツ)_/¯

ilyagr replied:
Ah, I see.

I tried `**expr.get(...).unwrap().clone()`, and it didn't work. The error is:

```
error[E0614]: type `std::option::Option<backend::CommitId>` cannot be dereferenced
    --> lib/src/git.rs:1350:21
     |
1350 | /                     **qualified_remote_refs_expected_locations
1351 | |                         .get(refname)
1352 | |                         .expect("Push is trying to move a ref it wasn't asked to move")
1353 | |                         .clone()
     | |________________________________^
```

But `(**expr.get(...).unwrap()).clone()` works. Let me know if you know a way to understand this.

necauqua replied:
Oh.. right :joy: 
Well `(*expr).clone()` is obvious and you understood it, we just turn &&T into &T and call a method that expects &T on it.

With `(**expr).clone()` I said it above, since we're calling a method that expects &T and we give it T it get's autoreferenced to &T, effectively turning our code into `(&**expr).clone()`, where `&*` cancel out and we get to the first option.
This "cancelling out" happens before ownership is considered I guess is the way to understand why it doesn't complain about that.

Rust auto(de)ref is joy, for example there's [this](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md)

ilyagr replied:
Oh, I think I got it. `**expr.get(...).unwrap().clone()` means `**(expr.get(...).unwrap().clone())`. `expr.get(...).unwrap()` is of type `&&Option<CommitId>`. `that.clone()` is of type `&Option<CommitId>`. Dereferencing that once gives `Option<CommitId>`, and the second dereference causes that error.

If there wasn't a type error, the dereference still wouldn't work as rustc would complain that `Option<CommitId>` is not Copy. This is what you get on `*expr.get(...).unwrap().clone()`.

ilyagr replied:
> Rust auto(de)ref is joy, for example there's [this](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md)

Yeah, that's why I spent an hour on this. I kept trying various combinations of `.as_ref()`, `.as_deref()`, 🤯

necauqua replied:
> what's going on in the former case

oh lol I thought you needed me to expand on the `(**expr).clone()` thing, well you figured it out ¯\\\_(ツ)_/¯

ilyagr commented on lib/src/git.rs:
```diff
@@ -1208,8 +1208,16 @@ pub enum GitPushError {
+    #[error("The push conflicts with changes made on the remote (it is not fast-forwardable).")]
     NotFastForward,
+    #[error("Refusing to push a branch that unexpectedly moved on the remote. Affected refs: {:?}",
+    .0
+                .iter()
+                .sorted()
+                .map(|option| option.clone().unwrap_or_default())
+                .collect_vec()
```
🤣 Yeah, I'm not sure what my plans are here, it's likely this part will change more significantly.

ilyagr replied:
Also, thanks for the idea, `function(.0)` probably does look a bit better (assuming it works).

yuja commented on lib/src/git.rs:
```diff
@@ -1324,66 +1331,166 @@ fn push_refs(
+                    // Conveniently, this also means that the push will be
+                    // successful **if and only if** `jj git fetch` is a no-op
+                    // for this branch (and, in particular, would not create a
+                    // conflict).
```
Out of curiosity, have you considered actually using our `merge_ref_targets()` logic? I think that's doable with some sanity check for unknown commit ids. Will it simplify the push negotiation?

ilyagr replied:
I thought about it, and then decided to keep this equivalence inside tests.

I think the behavior will be easier to debug and explain if it's written out. For example, this way we can have tracing messages explain the different situations.

I could add an assert.

yuja replied:
I would be happy if we can remove lengthy comment by using common ref-merge helper, but I don't have a strong opinion. The current implementation is also simple (except that we have to learn git2 API a bit more.)

martinvonz replied:
> I would be happy if we can remove lengthy comment by using common ref-merge helper, but I don't have a strong opinion.

+1 (both that it would be nice to avoid the long comment, and that I don't feel strongly)

ilyagr replied:
I'm not sure what I think at this point. As I alluded to, I do use the common ref-merge helper in tests, but I'm not sure how much of that (or the long comment) is actually helpful and how much is more of a historical artifact of me figuring out how things should work.

yuja replied:
Just for reference, this is my attempt to switch to be using merge_ref_targets(). Tests passed, but there might be some bugs.
```rust
                let oid_to_target = |oid: git2::Oid| {
                    let id = (!oid.is_zero()).then(|| CommitId::from_bytes(oid.as_bytes()));
                    RefTarget::resolved(id)
                };
                let remote_target = oid_to_target(update.src());
                if remote_target.added_ids().all(|id| repo.index().has_id(id)) {
                    // remote known, simulate merge
                    let base_target = RefTarget::resolved(expected_location.cloned());
                    let local_target = oid_to_target(update.dst());
                    let merged_target = refs::merge_ref_targets(
                        repo.index(),
                        &remote_target,
                        &base_target,
                        &local_target,
                    );
                    if merged_target != local_target {
                        failed_push_negotiations_ref.push(dst_refname.to_string());
                    }
                } else {
                    // remote unknown, must fetch
                    failed_push_negotiations_ref.push(dst_refname.to_string());
                }
```

ilyagr replied:
Thanks, this is close to what I was imagining and what I do in tests.

I still don't know what I think, but I don't consider this a blocker, my feelings about this are also not that strong. If I don't have any more concrete thoughts before we're done with all the other issues, I'll just use your piece of code and we can call it a day.

ilyagr replied:
I'm now thinking that if I change this to Yuya's way (which I'm still leaning towards), I'll add a comment and/or rework tests to make it clear what kind of fast-forwards are allowed (the logic I currently have written out explicitly). I just need to try it and see how it goes.

ilyagr replied:
I have a version of this, see the diff from the previous version here: https://github.com/ilyagr/jj/commit/fwl-mergerefs. It's longer than Yuya's since I again tried to explain both ways to understand the logic, this time more in code than in comments. The tests got a lot shorter, though.

I'll probably switch to it, or perhaps try to simplify it some more, but let me know if you have any thoughts.

yuja commented on lib/src/git.rs:
```diff
@@ -1229,14 +1230,11 @@ fn render_ref_in_unexpected_location_details(refs: &[String]) -> String {
     pub qualified_name: String,
     pub expected_current_target: Option<CommitId>,
-    // Short-term TODO: Delete `force`
-    pub force: bool,
```
I'm not pretty sure, but we might still need a "force" push for some integrations like gerrit.

@thoughtpolice do you have any idea?

ilyagr replied:
The plan is to always force push and do all the safety checks ourselves.

yuja replied:
> We end up always force pushing and doing all the safety checks ourselves.

Yeah, my concern was we might need a real "force" push (that doesn't respect remote refs at all) to implement some forge integrations, but I don't know. If there isn't an immediate need, we can restore it later.

ilyagr replied:
It's a good point that stranger git servers might send strange information to the push negotiation function or expect it to succeed in weird cases. We should hopefully be able to deal with it on a case-by-case basis and adjust the logic inside the push negotiation accordingly.

I don't know anybody I could ask to foresee this, perhaps Austin will have an idea. We could ask on the Git Discord, but I'm not sure how to phrase the question.

yuja commented on cli/src/commands/git.rs:
```diff
@@ -1015,7 +1015,7 @@ fn cmd_git_push(
         GitPushError::InternalGitError(err) => map_git_error(err),
         GitPushError::NotFastForward => user_error_with_hint(
-            "The push conflicts with changes made on the remote (it is not fast-forwardable).",
+            err.to_string(),
```
nit: can be just `err` (to preserve the source error chain)

ilyagr replied:
Good point. I didn't know how this error chain magic works; I should look at it.

I'm actually thinking of refactoring this a bit, making this less relevant. I'd like to get rid of the "NotFastForward refactor" commit, put the new error rendering logic here, and thus either get rid of the `render_unexpected_ref_details` function (or whatever I called it) or move it into the CLI.

ilyagr replied:
I now did refactor this in a different direction (moved the logic to the CLI). I can undo or change it some more if you prefer.

yuja commented on lib/src/git.rs:
```diff
@@ -1319,16 +1322,45 @@ fn push_refs(
+                .expect("Expect reference name to be valid UTF-8");
+            let expected_location = (*qualified_remote_refs_expected_locations
+                .get(dst_refname)
+                .expect("Push is trying to move a ref it wasn't asked to move"))
```
nit: suppose these data come from remote, it's probably better to return error instead of panicking.

yuja replied:
Never mind. It appears that the `updates` list is created from `push->specs` (which I think is local data.)

ilyagr replied:
I am less clear about the non-utf8 case, but I'm assuming some other part of jj would have panicked earlier if there was a non-utf8 ref floating around.

yuja replied:
The refspec is constructed by jj, so it should be valid UTF-8 I suppose.

yuja reviewed (commented):
I haven't looked through the changes yet. I'll review this PR hopefully in this weekend. Thanks for working on this!

ilyagr commented:
Thanks, there is no rush at all.

yuja commented on cli/tests/test_git_push.rs:
```diff
@@ -238,6 +238,118 @@ fn test_git_push_not_fast_forward() {
+        .jj_cmd(&origin_path, &["branch", "list", "--all"])
+        .assert()
+        .success();
+    insta::assert_snapshot!(get_stdout_string(&assert), @r###"
```
nit: `.jj_cmd_success()`, `_ok()`, or `_failure()` can be used instead of `.assert()`s.

I've also sent PR that adds `get_branch_output()` helper.
https://github.com/martinvonz/jj/pull/3715

ilyagr replied:
I was too lazy to fix this in all commits, but it should be fixed by the main "force-with-lease" commit.

yuja commented on cli/tests/test_git_push.rs:
```diff
@@ -350,6 +350,41 @@ fn test_git_push_deletion_unexpectedly_moved() {
+    let assert = test_env
+        .jj_cmd(&workspace_root, &["git", "push"])
+        .assert()
+        .failure();
```
nit: same here. `.jj_cmd_failure()`.

yuja commented on lib/src/git.rs:
```diff
@@ -1226,9 +1226,8 @@ pub struct GitBranchPushTargets {
-    // TODO: We want this to be a `current_target: Option<CommitId>` for the expected current
-    // commit on the remote. It's a blunt "force" option instead until git2-rs supports the
-    // "push negotiation" callback (https://github.com/rust-lang/git2-rs/issues/733).
+    pub expected_current_target: Option<CommitId>,
```
nit: can you add doc comment that says `None` doesn't mean "no expectation", but "expects absent". That will probably help when we have to add real force push (i.e. `expected: Any`.)

yuja commented on lib/src/git.rs:
```diff
@@ -1275,10 +1275,13 @@ pub fn push_updates(
-        qualified_remote_refs.push(update.qualified_name.as_str());
+        qualified_remote_refs_expected_locations.insert(
+            update.qualified_name.as_str(),
+            &update.expected_current_target,
```
nit: `.as_ref()` here will help simplify the code (see the comment below)

yuja commented on lib/src/git.rs:
```diff
@@ -1319,16 +1322,45 @@ fn push_refs(
+            let expected_location = (*qualified_remote_refs_expected_locations
+                .get(dst_refname)
+                .expect("Push is trying to move a ref it wasn't asked to move"))
+            .clone();
```
nit: `.clone()` isn't needed (and `Option<&CommitId>` is easier to deal with than `&Option<CommitId>`.)

yuja commented on lib/src/git.rs:
```diff
@@ -1319,16 +1322,45 @@ fn push_refs(
         }
     })?;
-    let mut remaining_remote_refs: HashSet<_> = qualified_remote_refs.iter().copied().collect();
+    let mut remaining_remote_refs_expected_locations =
```
nit: maybe better to leave `remaining_remote_refs` as `HashSet<_>`? We have to create an owned set or map anyway to get around mutable borrows.

yuja commented on lib/src/git.rs:
```diff
@@ -1218,6 +1222,10 @@ pub enum GitPushError {
 }
 
+fn render_ref_in_unexpected_location_details(refs: &[String]) -> String {
+    format!("Affected refs: {:?}", refs.iter().sorted().collect_vec())
```
nit: comma-separated list is probably better as an error message. `.join(", ")`

yuja commented on lib/src/git.rs:
```diff
@@ -1324,66 +1331,166 @@ fn push_refs(
-                GitPushError::NotFastForward
+    let mut failed_push_negotiations = vec![];
+    let push_result = {
+        let failed_push_negotiations_ref = &mut failed_push_negotiations;
```
nit: no need to create `&mut ` reference

ilyagr replied:
Good point! I'm quite confused by what the compiler does and does not do automatically; I thought this fixed an error but I guess something else I did actually fixed it.

yuja commented on lib/src/git.rs:
```diff
@@ -1324,66 +1331,166 @@ fn push_refs(
+        // information. See https://github.com/rust-lang/git2-rs/issues/1042.
+        assert!(push_result.is_err());
+        Err(GitPushError::RefInUnexpectedLocation(
+            failed_push_negotiations,
```
nit: sort here to stabilize error representation?

yuja commented on lib/tests/test_git.rs:
```diff
@@ -19,6 +19,7 @@ use std::{fs, iter, thread};
 
 use assert_matches::assert_matches;
 use git2::Oid;
+use insta::{assert_debug_snapshot, assert_snapshot};
```
nit: we usually spell `insta::assert_*()` at call sites (I don't know why, though.)

martinvonz replied:
FWIW, I wouldn't mind a PR that consistently imports the symbols.

ilyagr replied:
I'm OK with either convention. Explicit `insta::` might remind new people this is related to `cargo insta`.

yuja replied:
I don't mind either, but I think most of the insta assertions in `test_git.rs` can be removed.

yuja commented on lib/tests/test_git.rs:
```diff
@@ -2781,6 +2797,381 @@ fn test_push_branches_not_fast_forward_with_force() {
+            ),
+        ],
+    )
+    "###
```
Suppose these assertions are sanity checks, it's probably better to just `assert!(...has_conflicts())`. Insta snapshots are verbose.

yuja commented on lib/tests/test_git.rs:
```diff
@@ -2781,6 +2797,381 @@ fn test_push_branches_not_fast_forward_with_force() {
+    "###
+    );
+    let result = attempt_push_expecting_sideways(None);
+    assert_snapshot!(format!("{result:?}"), @r###"Err(RefInUnexpectedLocation(["refs/heads/main"]))"###);
```
nit: it can be `insta::assert_debug_snapshot!()`, but I would probably use `assert_matches!(.., Err(RefInUnexpectedLocation(_)))` since the failed ref names aren't important.

ilyagr commented on lib/src/git.rs:
```diff
@@ -1319,42 +1332,181 @@ fn push_refs(
+                "We allow the push of {dst_refname} to {dst}, even though it is unexpectedly at \
+                 {src} on the server rather than the expected {expected}. The desired location is \
+                 a descendant of the actual location, and the actual location is a descendant of \
+                 the expected location.",
```
Yes, I can do that; it makes the function a bit less messy. I haven't really changed the tracing; I think that printing those messages on `--debug` is fine, but we can also change it later.

yuja replied:
Tracing can be moved to caller. Suppose tracing output is for debugging, I think short easy-to-grep message is preferred.
```rust
let classification = allow_push(..);
tracing::info!(classification?);
```

ilyagr replied:
I did move it to the caller. I am not sure whether you mean that caller's caller or if you looked at an old version or if I misunderstood.

I'm still thinking of `--debug` partially as `--verbose`. I thought most people would not notice the more complicated parts of these push safety rules, but the ones that would care or be surprised by what `jj` did there could get quite confused by them. So, I thought I should explain them in detail, both for people who are reading the code and for people who run `--debug`.

If you really think that's unhelpful, we could remove some or all of it, I suppose.

Aside: I also had a version where the PushAllowReason had a method for printing the debug method, but I later decided that's overkill.

ilyagr commented on lib/src/git.rs:
```diff
@@ -1246,11 +1247,17 @@ pub fn push_branches(
         .collect_vec();
-    push_updates(git_repo, remote_name, &ref_updates, callbacks)?;
+    push_updates(
+        mut_repo.base_repo().as_ref(),
```
Done, thanks! Maybe I should have just used &Index, on second thought, but &dyn Repo should be fine too.

ilyagr commented:
This should be ready for review again.

Apart from addressing the comments, I added a commit for adjusting `jj git push`'s user-facing messages. I also removed the refactor of NotFastForward, since we delete it anyway.

yuja commented on lib/src/git.rs:
```diff
@@ -1319,12 +1326,42 @@ fn push_refs(
+                .expect("Push is trying to move a ref it wasn't asked to move");
+            let expected_location_oid = expected_remote_location
+                .map(|id| git2::Oid::from_bytes(id.as_bytes()).unwrap())
+                .unwrap_or(git2::Oid::zero());
```
nit: conversion to `git2::Oid` isn't needed anymore (and rewritten by the subsequent commit.)

yuja commented on cli/src/commands/git.rs:
```diff
@@ -1019,6 +1019,15 @@ fn cmd_git_push(
+            format!(
+                "Refusing to push a branch that unexpectedly moved on the remote. Affected refs: \
+                 {}",
+                refs.iter().sorted().join(", ")
```
nit: just `.join(", ")`

The list should be sorted.

yuja commented on lib/src/git.rs:
```diff
@@ -1330,65 +1335,191 @@ fn push_refs(
+
+                let location_to_string = |loc: Option<&CommitId>| {
+                    loc.map(|id| id.hex()).unwrap_or("(missing)".to_string())
+                };
```
nit: just use debug format?

I'm not strongly against adding long tracing message, but it doesn't have to be overly user-friendly. I think the debug formatting of `Option<&CommidId>` is pretty okay.

ilyagr replied:
You're probably right.

I was trying to come up with something that distinguishes "no expectation" from "expected to not exist", as you pointed out in one of your old comments. On second thought, "(missing)" is not much better than None in this regard. If I don't think of something better, I'll probably just switch to debug formatting.

yuja commented on lib/src/git.rs:
```diff
@@ -1330,65 +1335,191 @@ fn push_refs(
+    if actual_remote_location
+        .map(|id| !index.has_id(id))
+        .unwrap_or(false)
+    {
```
nit: you can use `.map_or()` to make it one line.

btw, I would write it as `!action_remote_location.map_or(true, ..)` to clarify that `None` is considered truthy in this context.

yuja commented on lib/src/git.rs:
```diff
@@ -1330,65 +1335,191 @@ fn push_refs(
+                    expected_remote_location.unwrap(),
+                    actual_remote_location.as_ref().unwrap()
+                )
+        );
```
nit: I don't think these debug assertions help. Maybe this can be reorganized as follows in order to make it clear that `} else {` means fast-forward?

```rust
if merge_ref_targets(...) == local_target { // means it is fast forward
    if remote_target == local_target {  // special case of fast-forward
        UnexpectedNoop
    } else {
        ExceptionalFastForward
    }
} else {
    Err  // not fast-forward
}
```

ilyagr replied:
I cleaned it up a bit more. I still feel conflicted about removing the "expected location is the ancestor of both other locations" assertion/explanation/verification.

ilyagr replied:
It was mainly because I found the logic confusing. but I'll trust your judgement that it didn't help.

yuja commented on lib/src/git.rs:
```diff
@@ -1420,12 +1417,7 @@ fn push_refs(
-                }
-                _ => GitPushError::InternalGitError(err),
-            })
+            .map_err(GitPushError::InternalGitError)
```
nit: `.map_err(..)` can be removed at all. (handled later by `?`)

ilyagr replied:
Nice!

yuja commented on cli/src/commands/git.rs:
```diff
@@ -964,21 +965,18 @@ fn cmd_git_push(
+                // possibly "Move branch ... sideways (X forward, Y back)".
+                writeln!(
+                    ui.status(),
+                    "  Move {dir} branch {branch_name} from {old} to {new}",
```
nit: it's probably better to map `dir` to `Forward|Backward|Sideways` enum.

We should avoid constructing English sentence in this way.

ilyagr replied:
I'm not sure about changing it now; this code will likely be updated a few times in the relatively near future anyway. I think your suggestion will happen naturally if we want to print any more details in the future, but is it worth the extra 10 lines now?

yuja replied:
I personally wouldn't mind adding two more match arms there, but I agree it can be addressed later. I think it's also good to split the last commit to new PR.

FWIW, I would use something like `{branch} (added|deleted|moved forward|...): {old} -> {new}` to keep the formatting code small. I think `jj git fetch` uses a similar formatting.

ilyagr replied:
>  I think it's also good to split the last commit to new PR.

Sure, I'll do that. I rushed the commit a bit because I don't like printing misleading "Force...", but we have more than a week until the next release, which should be plenty of time.

> FWIW, I would use something like {branch} (added|deleted|moved forward|...): {old} -> {new} to keep the formatting code small. I think jj git fetch uses a similar formatting.

This SGTM. For later, I guess it could become "{branch} moved forward {n}" or "{branch} moved sideways, back {n} and forward {m}".

ilyagr replied:
Yet another PR I hope to do before the next release is documenting this behavior, but it's OK if that slips to after the release.

ilyagr replied:
I haven't edited the commit yet, but I moved it to #3765

yuja reviewed (approved):
Generally looks good to me, thanks!

ilyagr commented:
Thank you Yuya for the review! I plan to merge this tomorrow if I don't hear otherwise. If I hear from you earlier, I might merge it then.
